File: backend/app/api/accounts.py
import imaplib

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.encryption import encrypt_value, decrypt_value
from app.database import get_db
from app.models.user import User
from app.models.email_account import EmailAccount, WatchedFolder
from app.schemas.email_account import (
    CreateAccountRequest, UpdateAccountRequest, AccountResponse,
    WatchFolderRequest, WatchedFolderResponse,
)
from app.api.deps import get_current_user

router = APIRouter(prefix="/api/v1/accounts", tags=["accounts"])


@router.get("", response_model=list[AccountResponse])
async def list_accounts(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(EmailAccount).where(EmailAccount.user_id == user.id))
    return result.scalars().all()


@router.post("", response_model=AccountResponse, status_code=status.HTTP_201_CREATED)
async def create_account(req: CreateAccountRequest, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = EmailAccount(
        user_id=user.id,
        name=req.name,
        imap_host=req.imap_host,
        imap_port=req.imap_port,
        imap_user=req.imap_user,
        imap_password_encrypted=encrypt_value(req.imap_password),
        use_ssl=req.use_ssl,
        polling_interval_sec=req.polling_interval_sec,
    )
    db.add(account)
    await db.commit()
    await db.refresh(account)
    return account


@router.patch("/{account_id}", response_model=AccountResponse)
async def update_account(account_id: int, req: UpdateAccountRequest, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = await db.get(EmailAccount, account_id)
    if not account or account.user_id != user.id:
        raise HTTPException(status_code=404, detail="Account not found")
    for field, value in req.model_dump(exclude_unset=True).items():
        if field == "imap_password" and value is not None:
            account.imap_password_encrypted = encrypt_value(value)
        elif hasattr(account, field):
            setattr(account, field, value)
    await db.commit()
    await db.refresh(account)
    return account


@router.delete("/{account_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_account(account_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = await db.get(EmailAccount, account_id)
    if not account or account.user_id != user.id:
        raise HTTPException(status_code=404, detail="Account not found")
    await db.delete(account)
    await db.commit()


@router.post("/{account_id}/test")
async def test_connection(account_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = await db.get(EmailAccount, account_id)
    if not account or account.user_id != user.id:
        raise HTTPException(status_code=404, detail="Account not found")
    try:
        password = decrypt_value(account.imap_password_encrypted)
        if account.use_ssl:
            mail = imaplib.IMAP4_SSL(account.imap_host, account.imap_port)
        else:
            mail = imaplib.IMAP4(account.imap_host, account.imap_port)
        mail.login(account.imap_user, password)
        mail.logout()
        return {"success": True}
    except Exception as e:
        return {"success": False, "error": str(e)}


@router.get("/{account_id}/folders", response_model=list[str])
async def list_folders(account_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = await db.get(EmailAccount, account_id)
    if not account or account.user_id != user.id:
        raise HTTPException(status_code=404, detail="Account not found")
    try:
        password = decrypt_value(account.imap_password_encrypted)
        if account.use_ssl:
            mail = imaplib.IMAP4_SSL(account.imap_host, account.imap_port)
        else:
            mail = imaplib.IMAP4(account.imap_host, account.imap_port)
        mail.login(account.imap_user, password)
        _, folder_list = mail.list()
        mail.logout()
        folders = []
        for item in folder_list:
            decoded = item.decode() if isinstance(item, bytes) else item
            parts = decoded.rsplit('" "', 1)
            if len(parts) == 2:
                folders.append(parts[1].strip('"'))
            else:
                parts = decoded.rsplit(" ", 1)
                folders.append(parts[-1].strip('"'))
        return folders
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Failed to list folders: {e}")


# --- Watched Folders ---

@router.get("/{account_id}/folders/watched", response_model=list[WatchedFolderResponse])
async def list_watched(account_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = await db.get(EmailAccount, account_id)
    if not account or account.user_id != user.id:
        raise HTTPException(status_code=404, detail="Account not found")
    result = await db.execute(select(WatchedFolder).where(WatchedFolder.account_id == account_id))
    return result.scalars().all()


@router.post("/{account_id}/folders/watched", response_model=WatchedFolderResponse, status_code=status.HTTP_201_CREATED)
async def add_watched(account_id: int, req: WatchFolderRequest, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = await db.get(EmailAccount, account_id)
    if not account or account.user_id != user.id:
        raise HTTPException(status_code=404, detail="Account not found")
    folder = WatchedFolder(account_id=account_id, folder_path=req.folder_path)
    db.add(folder)
    await db.commit()
    await db.refresh(folder)
    return folder


@router.delete("/{account_id}/folders/watched/{folder_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_watched(account_id: int, folder_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    account = await db.get(EmailAccount, account_id)
    if not account or account.user_id != user.id:
        raise HTTPException(status_code=404, detail="Account not found")
    folder = await db.get(WatchedFolder, folder_id)
    if not folder or folder.account_id != account_id:
        raise HTTPException(status_code=404, detail="Folder not found")
    await db.delete(folder)
    await db.commit()

File: frontend/src/stores/accounts.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/api/client'

export interface EmailAccount {
  id: number
  name: string
  imap_host: string
  imap_port: number
  imap_user: string
  use_ssl: boolean
  polling_interval_sec: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface CreateAccountPayload {
  name: string
  imap_host: string
  imap_port: number
  imap_user: string
  imap_password: string
  use_ssl: boolean
  polling_interval_sec: number
}

export interface UpdateAccountPayload {
  name?: string
  imap_host?: string
  imap_port?: number
  imap_user?: string
  imap_password?: string
  use_ssl?: boolean
  polling_interval_sec?: number
  is_active?: boolean
}

export interface IMAPFolder {
  name: string
}

export interface WatchedFolder {
  id: number
  folder_path: string
  last_seen_uid: number
}

export const useAccountsStore = defineStore('accounts', () => {
  const accounts = ref<EmailAccount[]>([])
  const loading = ref(false)

  async function fetchAccounts() {
    loading.value = true
    try {
      const res = await api.get('/accounts')
      accounts.value = res.data
    } finally {
      loading.value = false
    }
  }

  async function createAccount(data: CreateAccountPayload): Promise<EmailAccount> {
    const res = await api.post('/accounts', data)
    accounts.value.push(res.data)
    return res.data
  }

  async function updateAccount(id: number, data: UpdateAccountPayload): Promise<EmailAccount> {
    const res = await api.patch(`/accounts/${id}`, data)
    const idx = accounts.value.findIndex((a) => a.id === id)
    if (idx !== -1) accounts.value[idx] = res.data
    return res.data
  }

  async function deleteAccount(id: number): Promise<void> {
    await api.delete(`/accounts/${id}`)
    accounts.value = accounts.value.filter((a) => a.id !== id)
  }

  async function testConnection(id: number): Promise<{ success: boolean; message: string }> {
    const res = await api.post(`/accounts/${id}/test`)
    return res.data
  }

  async function fetchFolders(id: number): Promise<IMAPFolder[]> {
    const res = await api.get(`/accounts/${id}/folders`)
    return (res.data as string[]).map((name) => ({ name }))
  }

  async function fetchWatchedFolders(id: number): Promise<WatchedFolder[]> {
    const res = await api.get(`/accounts/${id}/folders/watched`)
    return res.data
  }

  async function addWatchedFolder(id: number, folderPath: string): Promise<WatchedFolder> {
    const res = await api.post(`/accounts/${id}/folders/watched`, { folder_path: folderPath })
    return res.data
  }

  async function removeWatchedFolder(id: number, folderId: number): Promise<void> {
    await api.delete(`/accounts/${id}/folders/watched/${folderId}`)
  }

  return {
    accounts,
    loading,
    fetchAccounts,
    createAccount,
    updateAccount,
    deleteAccount,
    testConnection,
    fetchFolders,
    fetchWatchedFolders,
    addWatchedFolder,
    removeWatchedFolder,
  }
})
